src/stream_io.rs

```diff
@@ -13,10 +11,9 @@ pub unsafe extern "C" fn write_stream<T: Write>(writer: *mut c_void,
                                                 n: c_int)
                                                 -> c_int {
     let bytes = slice::from_raw_parts(c as *const u8, n as usize);
-    if let Ok(_) = (*(writer as *mut T)).write_all(bytes) {
-        return 0;
-    } else {
-        return 1;
+    match (*(writer as *mut T)).write_all(bytes) {
+        Ok(_) => return 0,
+        Err(e) => return e.raw_os_error().unwrap_or(1) as c_int,
```

Comment:
1 is not a safe sentinel value here. For example, on Linux, error 1 is EPERM, or "Permission Denied." Reporting all non-OS `io::Error` values as EPERM would be very confusing. Finding a cross-platform safe sentinel value is probably more difficult than just separating the success flag from the error code.